### PR TITLE
fix issue 'QFutureWatcher::connect: connecting after calling setFutur…

### PR DESCRIPTION
**Pull request by kemen209**
## Description

Fix warning: QFutureWatcher::connect: connecting after calling setFuture() is likely to produce race.

The right step to use a QFutureWatcher should be as follow:
```
1. Create a QFutureWatcher instance
2. Connect to signals if necessary
3. Get a future by schdule a background task.
4. Set future to QFutureWatcher.
```

### CODE REVIEW

File: src/3d/terrain/qgsdemterraintileloader_p.cpp [hunk 1]
```diff
@@ -229,16 +229,18 @@ int QgsDemHeightMapGenerator::render( const QgsChunkNodeId &nodeId )
   jd.tileId = nodeId;
   jd.extent = extent;
   jd.timer.start();
+
```
Comment (nyalldawson) [inline src/3d/terrain/qgsdemterraintileloader_p.cpp:232]:
```suggestion

```

File: src/3d/terrain/qgsdemterraintileloader_p.cpp [hunk 2]
```diff
@@ -229,16 +229,18 @@ int QgsDemHeightMapGenerator::render( const QgsChunkNodeId &nodeId )
   jd.tileId = nodeId;
   jd.extent = extent;
   jd.timer.start();
+  
+  QFutureWatcher<QByteArray> *fw = new QFutureWatcher<QByteArray>( nullptr );
+  connect( fw, &QFutureWatcher<QByteArray>::finished, this, &QgsDemHeightMapGenerator::onFutureFinished );
+  connect( fw, &QFutureWatcher<QByteArray>::finished, fw, &QObject::deleteLater );
+
```
Comment (nyalldawson) [inline src/3d/terrain/qgsdemterraintileloader_p.cpp:236]:
```suggestion

```

File: src/analysis/vector/geometry_checker/qgsgeometrychecker.cpp [hunk 3]
```diff
@@ -82,15 +82,16 @@ QFuture<void> QgsGeometryChecker::execute( int *totalSteps )
...
     }
   }
-
-  QFuture<void> future = QtConcurrent::map( mChecks, RunCheckWrapper( this ) );
-
-  QFutureWatcher<void> *watcher = new QFutureWatcher<void>();
-  watcher->setFuture( future );
   QTimer *timer = new QTimer();
   connect( timer, &QTimer::timeout, this, &QgsGeometryChecker::emitProgressValue );
+
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsgeometrychecker.cpp:87]:
```suggestion

```

File: src/analysis/vector/geometry_checker/qgsgeometrychecker.cpp [hunk 4]
```diff
@@ -82,15 +82,16 @@ QFuture<void> QgsGeometryChecker::execute( int *totalSteps )
...
   QTimer *timer = new QTimer();
   connect( timer, &QTimer::timeout, this, &QgsGeometryChecker::emitProgressValue );
+  
+  QFutureWatcher<void> *watcher = new QFutureWatcher<void>();
   connect( watcher, &QFutureWatcherBase::finished, timer, &QObject::deleteLater );
   connect( watcher, &QFutureWatcherBase::finished, watcher, &QObject::deleteLater );
+
+  QFuture<void> future = QtConcurrent::map( mChecks, RunCheckWrapper( this ) );
+  watcher->setFuture( future );
+
```
Comment (nyalldawson) [inline src/analysis/vector/geometry_checker/qgsgeometrychecker.cpp:94]:
```suggestion

```

File: src/app/qgsgeometryvalidationservice.cpp [hunk 5]
```diff
@@ -456,6 +456,37 @@ void QgsGeometryValidationService::triggerTopologyChecks( QgsVectorLayer *layer,
 
   mLayerChecks[layer].topologyCheckFeedbacks = feedbacks.values();
 
+  QFutureWatcher<void> *futureWatcher = new QFutureWatcher<void>();
+
```
Comment (nyalldawson) [inline src/app/qgsgeometryvalidationservice.cpp:460]:
```suggestion

```

**Review by nyalldawson — APPROVED** [review]

**Comment by kemen209** [comment]
> Check the details for the failing indentation test here: https://github.com/qgis/QGIS/pull/45569/checks?check_run_id=3940330123
> 
> There's a description in the test output which details what you need to do 👍

 I haven't notice the indentation test 😓 , thanks for pointing out. 👍

**Comment by kemen209** [comment]
> @kemen209
> 
> > But i do not have any easy way to do it now. So just keep it that way in this change.
> 
> I wonder if this could be handled by adding connections to BOTH QgsMapRendererParallelJob::renderLayersFinished/ QgsMapRendererParallelJob::renderingFinished at the time mFutureWatcher is constructed, and then use a bool member variables to set whether renderLayersFinished and/or renderingFinished should just exit immediately. Then we could avoid this disconnect/reconnect here and instead just set the bool values accordingly... Thoughts?

That's my init thougths too. But after i rethought about the purpose of this disconnect the renderLayersFinished signal and connect a new renderingFinished signal, i came out a new idea: 

Remove these disconnect-and-connect code entirely, keep the original signal, let it call renderLayersFinished directly, as such:
```
diff --git a/src/core/maprenderer/qgsmaprendererparalleljob.cpp b/src/core/maprenderer/qgsmaprendererparalleljob.cpp
index 79251541880..34e622b4898 100644
--- a/src/core/maprenderer/qgsmaprendererparalleljob.cpp
+++ b/src/core/maprenderer/qgsmaprendererparalleljob.cpp
@@ -131,12 +131,6 @@ void QgsMapRendererParallelJob::cancelWithoutBlocking()
     if ( job.renderer && job.renderer->feedback() )
       job.renderer->feedback()->cancel();
   }
-
-  if ( mStatus == RenderingLayers )
-  {
-    disconnect( &mFutureWatcher, &QFutureWatcher<void>::finished, this, &QgsMapRendererParallelJob::renderLayersFinished );
-    connect( &mFutureWatcher, &QFutureWatcher<void>::finished, this, &QgsMapRendererParallelJob::renderingFinished );
-  }
 }
```

(1) The purpose of these disconnect-and-connect code is try to bypass the call to renderLayersFinished and directly call on renderingFinished when in 'RenderingLayers' stage.

(2) After my change, the renderLayersFinished will call renderingFinished directly since 'mLabelJob' is been set to 'renderingStopped'. The only difference betwwen the change and the orignal is that few more codes will be excuted after the change:
```
void QgsMapRendererParallelJob::renderLayersFinished()
{
  Q_ASSERT( mStatus == RenderingLayers );

  for ( const LayerRenderJob &job : mLayerJobs )
  {
    if ( !job.errors.isEmpty() )
    {
      mErrors.append( Error( job.layerId, job.errors.join( ',' ) ) );
    }
  }

  // compose final image for labeling
  if ( mSecondPassLayerJobs.empty() )
  {
    mFinalImage = composeImage( mSettings, mLayerJobs, mLabelJob, mCache );
  }
```
But i think it's not a big deal, maybe a flag to not composeImage? And i'm confused about this composeImage, why 
here, shouldn't it be more properly in renderingFinished?

(3) There is another catch for cancelWithoutBlocking(), it didn't cancel 'mSecondPassLayerJobs', it will be carry out if it's not empty(take a look at renderingFinished), no matter whether mLayerJobs/mLabelJob been canceled or not. Since i not fully understand this 'mSecondPassLayerJobs', i will leave that to you.

Thanks.

**Comment by nyalldawson** [comment]
@kemen209 
> And i'm confused about this composeImage, why
here, shouldn't it be more properly in renderingFinished?

I suspect it's just to avoid to unwanted work of composing the map when the job has already been discarded and the results aren't needed anymore. (Composing is non-trivial, it comes with a cost)

> (3) There is another catch for cancelWithoutBlocking(), it didn't cancel 'mSecondPassLayerJobs', it will be carry out if it's not empty(take a look at renderingFinished), no matter whether mLayerJobs/mLabelJob been canceled or not. Since i not 
fully understand this 'mSecondPassLayerJobs', i will leave that to you.

Unfortunately it's not well understood code in general, and not in widespread use currently. I think there's likely a fair number of bugs in the second pass layer jobs handling in general

**Pull request merged**